### PR TITLE
standardize bookmarks

### PR DESCRIPTION
**Pull request by AngryBerryMS**
[AB#434653](https://dev.azure.com/ceapex/Engineering/_workitems/edit/434653)

### CODE REVIEW

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 1]
```diff
@@ -289,7 +289,35 @@ private static bool IsXrefSpec(JObject obj, JsonSchemaMap schemaMap, out SourceI
...
         {
             var siteUrl = _documentProvider.GetSiteUrl(file);
-            return !isRootLevel && uidCount > 1 ? UrlUtility.MergeUrl(siteUrl, "", $"#{Regex.Replace(uid, @"\W", "_")}") : siteUrl;
+            return !isRootLevel && uidCount > 1 ? UrlUtility.MergeUrl(siteUrl, "", $"#{GetStandardizedBookmark(uid)}") : siteUrl;
+        }
+
+        private static string GetStandardizedBookmark(string uid)
+        {
+#pragma warning disable CA1304 // Specify CultureInfo
+            uid = uid.ToLower();
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs:298]: remove "#pragma warning disable CA1304"
```suggestion
            uid = uid.ToLowerInvarient();
```

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 2]
```diff
@@ -289,7 +289,35 @@ private static bool IsXrefSpec(JObject obj, JsonSchemaMap schemaMap, out SourceI
...
+#pragma warning restore CA1304 // Specify CultureInfo
+
+            // Remove  ", %, \, ^ and '
+            uid = Regex.Replace(uid, @"[""'%^\\]", "");
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs:302]: use string replace like line 307

File: docs/specs/xref.yml [hunk 3]
```diff
@@ -2027,3 +2027,61 @@ outputs:
         {"uid": "c", "href": "https://docs.com/docs/c/c", "name": "Title from yaml header c"}
       ]
     }
+---
```
Comment (yufeih) [inline docs/specs/xref.yml:2030]: use xunit `InlineData` test to test the `GetStandardizedBookmark` helper method?

File: src/docfx/lib/schema/JsonSchemaTransformer.cs (hunk 1 above)
Comment (AngryBerryMS) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: Fixed.

File: src/docfx/lib/schema/JsonSchemaTransformer.cs (hunk 2 above)
Comment (AngryBerryMS) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: Done

File: docs/specs/xref.yml (hunk 3 above)
Comment (AngryBerryMS) [reply docs/specs/xref.yml]: Done.

File: src/docfx/lib/UrlUtility.cs [hunk 4]
```diff
@@ -348,6 +348,31 @@ public static string RemoveLeadingHostName(string url, string hostName, bool rem
                 : $"/{path}";
         }
 
+        public static string StandardizeBookmark(string uid)
+        {
+            uid = uid.ToLowerInvariant();
```
Comment (live1206) [inline src/docfx/lib/UrlUtility.cs:353]: can be combined to L356 as well

**Review by live1206 — APPROVED** [review]

File: src/docfx/lib/UrlUtility.cs [hunk 5]
```diff
@@ -348,6 +348,31 @@ public static string RemoveLeadingHostName(string url, string hostName, bool rem
...
+                Replace("{", "((").
+                Replace("}", "))");
+
+            uid = Regex.Replace(uid, @"[^a-zA-Z0-9()]", "-");
+
+            uid = Regex.Replace(uid, @"^-+", "");
+            uid = Regex.Replace(uid, @"-+$", "");
+            var bookmark = Regex.Replace(uid, @"-+", "-");
+
+            return bookmark.Length <= 100 ? bookmark : $"{bookmark.Substring(0, 100)}";
```
Comment (yufeih) [inline src/docfx/lib/UrlUtility.cs:373]: remove unneeded string interpolation
```suggestion
            return bookmark.Length <= 100 ? bookmark : bookmark.Substring(0, 100);
```

File: src/docfx/lib/UrlUtility.cs [hunk 6]
```diff
@@ -348,6 +348,31 @@ public static string RemoveLeadingHostName(string url, string hostName, bool rem
                 : $"/{path}";
         }
 
+        public static string StandardizeBookmark(string uid)
```
Comment (yufeih) [inline src/docfx/lib/UrlUtility.cs:351]: prefer more common terms like get bookmark

File: src/docfx/lib/UrlUtility.cs [hunk 7]
```diff
@@ -348,6 +348,31 @@ public static string RemoveLeadingHostName(string url, string hostName, bool rem
...
+        {
+            uid = uid.ToLowerInvariant();
+
+            uid =
```
Comment (yufeih) [inline src/docfx/lib/UrlUtility.cs:355]: so when we chain calls, the usual practice is to put `.` at the beginning:

```csharp
uid = uid
    .A
    .B
```

Comment (AngryBerryMS) [reply src/docfx/lib/UrlUtility.cs:353]: Fixed

File: src/docfx/lib/UrlUtility.cs (hunk 6 above)
Comment (AngryBerryMS) [reply src/docfx/lib/UrlUtility.cs:351]: Fixed.

File: src/docfx/lib/UrlUtility.cs (hunk 4 above)
Comment (AngryBerryMS) [reply src/docfx/lib/UrlUtility.cs]: Done

**Pull request merged**